reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,85 +83,116 @@ object StreamSubscriber {
     def stream(subscribe: F[Unit])(implicit ev: ApplicativeError[F, Throwable]): Stream[F, A] =
       Stream.bracket(subscribe)(_ => onFinalize) >> Stream
-        .eval(dequeue1)
+        .evalUnChunk(dequeue1)
         .repeat
         .rethrow
         .unNoneTerminate
   }
 
-  private[reactivestreams] def fsm[F[_], A](implicit F: Concurrent[F]): F[FSM[F, A]] = {
-    type Out = Either[Throwable, Option[A]]
+  private[reactivestreams] def fsm[F[_], A](
+      q: Queue[F, Either[Throwable, Option[A]]],
+      batchSize: Int
+  )(implicit F: Concurrent[F]): F[FSM[F, A]] = {
 
     sealed trait Input
     case class OnSubscribe(s: Subscription) extends Input
     case class OnNext(a: A) extends Input
     case class OnError(e: Throwable) extends Input
     case object OnComplete extends Input
     case object OnFinalize extends Input
-    case class OnDequeue(response: Deferred[F, Out]) extends Input
+    case object OnDequeue extends Input
 
     sealed trait State
     case object Uninitialized extends State
     case class Idle(sub: Subscription) extends State
-    case class RequestBeforeSubscription(req: Deferred[F, Out]) extends State
-    case class WaitingOnUpstream(sub: Subscription, elementRequest: Deferred[F, Out]) extends State
+    case class NextBatch(sub: Subscription) extends State
+    case class Receiving(sub: Subscription) extends State
+    case object RequestBeforeSubscription extends State
+    case class WaitingOnUpstream(sub: Subscription) extends State
     case object UpstreamCompletion extends State
     case object DownstreamCancellation extends State
     case class UpstreamError(err: Throwable) extends State
 
-    def step(in: Input): State => (State, F[Unit]) =
+    def step(in: Input, counter: Long): State => (State, F[Unit]) =
       in match {
         case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) => WaitingOnUpstream(s, req) -> F.delay(s.request(1))
-          case Uninitialized                  => Idle(s) -> F.unit
+          case RequestBeforeSubscription => WaitingOnUpstream(s) -> F.delay(s.request(batchSize))
+          case Uninitialized             => Idle(s) -> F.unit
+          case Idle(_)                   => Idle(s) -> F.delay(s.cancel)
           case o =>
             val err = new Error(s"received subscription in invalid state [$o]")
             o -> (F.delay(s.cancel) >> F.raiseError(err))
         }
         case OnNext(a) => {
-          case WaitingOnUpstream(s, r) => Idle(s) -> r.complete(a.some.asRight)
-          case DownstreamCancellation  => DownstreamCancellation -> F.unit
-          case o                       => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
+          case WaitingOnUpstream(s) if batchSize > 1   => Receiving(s) -> q.enqueue1(a.some.asRight)
+          case WaitingOnUpstream(s) if batchSize === 1 => Idle(s) -> q.enqueue1(a.some.asRight)
+          case Receiving(s) if counter % batchSize === 0 =>
```

Comment:
Waiting until we've received `batchSize` elements before requesting more would seem to result in potentially less throughput than if we requested as we dequeued each chunk. Requesting as we go would allow the producer to keep producing while the consumer is processing.

Maybe if we only emitted `batchSize` chunks, this would be desirable, but we currently dequeue whatever is available (which is also fine, I think, as consumers can always rechunk if they want a particular size).

Reply:
Alternatively, if we do want to produce fixed size chunks and not automatically prefetch by default, we could use an implementation similar to what is in upstream but maintaining a `batchSize` buffer of elements instead of a single one. Clients could then choose to prefetch using `Stream#prefetch`.